lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +433,811 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+            } else {
+              Queue.insert(Succ);
+            }
+          }
+        }
+      }
+
+      Block.End = End + Size;
+      CFG.insert(Block);
+    }
+  }
+
+  return CFG;
+}
+
+template<class FO>
+FunctionSummary
+CFEPAnalyzer<FO>::analyze(GlobalsArrayRef ABIRegisters,
+                          llvm::BasicBlock *Entry,
+                          llvm::raw_fd_ostream &OutputIBI,
+                          llvm::raw_fd_ostream &OutputAAWriter) {
+  using namespace llvm;
+
+  Value *RA = nullptr;
+  const auto *PCH = GCBI->programCounterHandler();
+  StructType *MetaAddressTy = MetaAddress::getStruct(&M);
+
+  // Run function boundaries detection, and retrieve the outlined function.
+  Function *OutFunc = createDisposableFunction(Entry);
+
+  // Initialize CFG of the function for the model.
+  auto CFG = collectDirectCFG(Entry, OutFunc);
+
+  // Identify the callee-saved registers and tell if a function returns
+  // properly, i.e., it jumps to the return address. To achieve this, we craft
+  // the IR by loading the PC, the SP, as well as the ABI registers CSVs
+  // respectively at function entry and end. We win when the subtraction between
+  // their entry and end values is zero.
+  IRBuilder<> Builder(&OutFunc->getEntryBlock().front());
+  auto *IntPtrTy = GCBI->spReg()->getType();
+  auto *IntTy = GCBI->spReg()->getType()->getElementType();
+
+  auto *SP = Builder.CreateLoad(GCBI->spReg());
+  auto *SPPtr = Builder.CreateIntToPtr(SP, IntPtrTy);
+  auto *GEPI = Builder.CreateGEP(IntTy, SPPtr, ConstantInt::get(IntTy, 0));
+
+  switch (GCBI->arch()) {
+  case Triple::arm: {
+    constexpr auto *LinkRegister = "r14";
+    RA = Builder.CreateLoad(M.getGlobalVariable(LinkRegister, true));
+    break;
+  }
+
+  case Triple::aarch64: {
+    constexpr auto *LinkRegister = "lr";
+    RA = Builder.CreateLoad(M.getGlobalVariable(LinkRegister, true));
+    break;
+  };
+
+  case Triple::mips:
+  case Triple::mipsel: {
+    constexpr auto *ReturnRegister = "ra";
+    RA = Builder.CreateLoad(M.getGlobalVariable(ReturnRegister, true));
+    break;
+  }
+
+  case Triple::systemz: {
+    constexpr auto *ReturnRegister = "r14";
+    RA = Builder.CreateLoad(M.getGlobalVariable(ReturnRegister, true));
+    break;
+  }
+
+  case Triple::x86:
+  case Triple::x86_64: {
+    RA = Builder.CreateLoad(SPPtr);
+    break;
+  }
+
+  default:
+    revng_abort("Unsupported architecture");
+  }
+
+  std::array<Value *, 4> DissectedPC = PCH->dissectJumpablePC(Builder,
+                                                              RA,
+                                                              GCBI->arch());
+
+  auto *PCI = MetaAddress::composeIntegerPC(Builder,
+                                            DissectedPC[0],
+                                            DissectedPC[1],
+                                            DissectedPC[2],
+                                            DissectedPC[3]);
+
+  SmallVector<Value *, 16> CSVI;
```

Comment:
Reading the code, at this point, I'm kind of losing track of what's going on.
The last comment told me the bird's eye view of what's going on, but I'd like more frequent comments.
For instance, you're creating `indirect_branch_info`. What is it? What's its purpose?
Also, consider outlining some self-contained code into support functions.

The function is veeeery long.